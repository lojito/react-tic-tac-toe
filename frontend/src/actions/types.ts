import {
  Categories,
  FirstPlayer,
  GameResult,
  ImageName,
  PlayingLevel,
  SquareLocation,
} from "../types";

export type CategoryAction = {
  type: "CHANGE_CATEGORY";
  category: Categories;
};

export type DisabledAction = {
  type: "CHANGE_DISABLED";
  disabled: boolean;
};

export type FirstAction = {
  type: "CHANGE_FIRST";
  first: FirstPlayer;
};

export type UserImageAction = {
  type: "CHANGE_USER_IMAGE";
  userImage: ImageName;
};

export type ComputerImageAction = {
  type: "CHANGE_COMPUTER_IMAGE";
  computerImage: ImageName;
};

export type PlayingLevelAction = {
  type: "CHANGE_PLAYING_LEVEL";
  level: PlayingLevel;
};

export type GameOverAction = {
  type: "CHANGE_GAME_OVER";
  over: boolean;
};

export type ResultAction = {
  type: "CHANGE_GAME_RESULT";
  result: GameResult;
};

export type WinnersSquaresAction = {
  type: "CHANGE_WINNERS_SQUARES";
  winners: SquareLocation[];
};

export type ActionUnknow = {
  type: "UNKNOWN";
};

export type Action =
  | CategoryAction
  | DisabledAction
  | FirstAction
  | UserImageAction
  | ComputerImageAction
  | PlayingLevelAction
  | GameOverAction
  | ResultAction
  | WinnersSquaresAction
  | ActionUnknow;
