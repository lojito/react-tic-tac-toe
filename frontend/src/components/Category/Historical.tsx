import React, { FC, useContext } from "react";
import { DictionaryContext } from "../../contexts/dictionary/DictionaryContext";
import categories from "../../data/json/categories.json";
import { Categories, CategoryId } from "../../types";
import "./Category.scss";

interface Props {
  categoryId: CategoryId;
}

const Historical: FC<Props> = ({ categoryId }) => {
  const { dictionary } = useContext(DictionaryContext);

  return (
    <div className="categories" data-testid="categories">
      {dictionary.CATEGORY}:{" "}
      {dictionary[(categories as Categories[])[categoryId].name]}
    </div>
  );
};

export default Historical;
