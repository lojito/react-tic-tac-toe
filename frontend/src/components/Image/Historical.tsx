import React, { FC, useContext } from "react";
import { DictionaryContext } from "../../contexts/dictionary/DictionaryContext";
import { ImageFolder, ImageName } from "../../types";
import "./Image.scss";

interface Props {
  folder: ImageFolder;
  userImage: ImageName;
  computerImage: ImageName;
}

const Historical: FC<Props> = ({ folder, userImage, computerImage }) => {
  const {
    dictionary: { IMAGE_USER_LABEL, IMAGE_COMPUTER_LABEL },
  } = useContext(DictionaryContext);
  const url = `${process.env.PUBLIC_URL}/images`;

  return (
    <div className="images historical" data-testid="images-historical">
      <div className="user-image">
        <span>{`${IMAGE_USER_LABEL}: `}</span>
        <img
          className="image-token"
          src={`${url}/${folder}/${userImage}.jpg`}
          alt="user"
        />
      </div>

      <div className="computer-image">
        <span>{`${IMAGE_COMPUTER_LABEL}: `}</span>
        <img
          className="image-token"
          src={`${url}/${folder}/${computerImage}.jpg`}
          alt="computer"
        />
      </div>
    </div>
  );
};

export default Historical;
