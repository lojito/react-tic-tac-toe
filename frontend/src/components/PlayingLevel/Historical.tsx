import React, { FC, useContext } from "react";
import { DictionaryContext } from "../../contexts/dictionary/DictionaryContext";
import { PlayingLevel } from "../../types";
import "./PlayingLevel.scss";

interface Props {
  level: PlayingLevel;
}

const Historical: FC<Props> = ({ level }) => {
  const {
    dictionary: {
      PLAYING_LEVEL_LABEL,
      PLAYING_LEVEL_EASY,
      PLAYING_LEVEL_SMART,
    },
  } = useContext(DictionaryContext);

  return (
    <div className="level" data-testid="level">
      <div>
        {PLAYING_LEVEL_LABEL}:{" "}
        {level === PlayingLevel.Easy ? PLAYING_LEVEL_EASY : PLAYING_LEVEL_SMART}
      </div>
    </div>
  );
};

export default Historical;
