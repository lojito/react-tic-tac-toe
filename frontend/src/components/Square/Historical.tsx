import React, { FC, memo } from "react";
import defaultImage from "../../assets/images/default.jpg";
import { SquareBoard } from "../../types";
import "./Square.scss";

const Square: FC<SquareBoard> = memo(
  ({ disabled, id, imagePath, className }) => {
    let url;
    if (imagePath === "default") {
      url = `url(${defaultImage}`;
    } else {
      url = `url(${process.env.PUBLIC_URL}/images/${imagePath}.jpg)`;
    }

    return (
      <button
        className={className}
        disabled
        id={id}
        style={{ backgroundImage: url }}
        aria-label="historical"
      />
    );
  }
);

export default Square;
