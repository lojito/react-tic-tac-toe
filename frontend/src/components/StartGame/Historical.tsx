import React, { FC, useContext } from "react";
import { DictionaryContext } from "../../contexts/dictionary/DictionaryContext";
import { FirstPlayer, Player } from "../../types";
import "./StartGame.scss";

interface Props {
  first: FirstPlayer;
}

const Historical: FC<Props> = ({ first }) => {
  const {
    dictionary: {
      GAME_STARTED_BY,
      GAME_STARTED_BY_COMPUTER,
      GAME_STARTED_BY_YOU,
    },
  } = useContext(DictionaryContext);

  return (
    <div className="start" data-testid="start">
      {GAME_STARTED_BY}:{" "}
      {first === Player.Computer
        ? GAME_STARTED_BY_COMPUTER
        : GAME_STARTED_BY_YOU}
    </div>
  );
};

export default Historical;
