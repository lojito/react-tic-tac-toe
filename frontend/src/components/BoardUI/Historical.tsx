import React, { FC, memo } from "react";
import {
  GameBoard,
  ImageFolder,
  ImageName,
  Player,
  SquareLocation,
} from "../../types";
import Square from "../Square/Historical";
import "./BoardUI.scss";

interface Props {
  folder: ImageFolder;
  computerImage: ImageName;
  userImage: ImageName;
  board: GameBoard;
  winners: SquareLocation[];
}

const Historical: FC<Props> = ({
  folder,
  userImage,
  computerImage,
  board,
  winners,
}) => {
  const boardJSX = board.map((player: Player, square: SquareLocation) => {
    const imagePath =
      player === Player.Computer
        ? `${folder}/${computerImage}`
        : player === Player.User
        ? `${folder}/${userImage}`
        : "default";
    const className = winners.indexOf(square) !== -1 ? "win" : "";

    return (
      <Square
        disabled
        id={square + ""}
        imagePath={imagePath}
        key={`square-${square}`}
        className={className}
      />
    );
  });

  return <div className="board">{boardJSX}</div>;
};

export default memo(Historical);
