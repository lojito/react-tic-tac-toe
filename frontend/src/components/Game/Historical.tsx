import { useMutation } from "@apollo/client";
import React, { FC, useContext, useEffect } from "react";
import { Link, useHistory } from "react-router-dom";
import { DictionaryContext } from "../../contexts/dictionary/DictionaryContext";
import categories from "../../data/json/categories.json";
import DELETE_GAME from "../../mutations/deleteGame";
import GET_GAMES from "../../queries/getGames";
import { Categories, DBGame, GameResult } from "../../types";
import BoardUI from "../BoardUI/Historical";
import Category from "../Category/Historical";
import Images from "../Image/Historical";
import PlayingLevel from "../PlayingLevel/Historical";
import StartGame from "../StartGame/Historical";
import "./Game.scss";

interface Props {
  token: string;
  game: DBGame;
  displayBoard: boolean;
}

const Historical: FC<Props> = ({ game, displayBoard, token }) => {
  const history = useHistory();
  const {
    dictionary: {
      MESSAGE_LOST,
      MESSAGE_WON,
      MESSAGE_DRAW,
      BUTTON_SEE_GAME,
      BUTTON_BACK,
      BUTTON_DELETE_GAME,
      GAME_DELETING,
      GAME_DELETE_ERROR,
    },
  } = useContext(DictionaryContext);

  const context = { headers: { Authorization: `Bearer ${token}` } };
  const [deleteGame, { data, loading, error }] = useMutation(DELETE_GAME, {
    context,
    refetchQueries: [{ query: GET_GAMES, context }],
  });

  const {
    categoryId,
    computerImage,
    userImage,
    first,
    level,
    board,
    winners,
    result,
  } = game;

  const id = game._id!;

  const handleDeleteGame = async (gameId: string) => {
    deleteGame({ variables: { gameId: gameId } });
  };

  useEffect(() => {
    if (data) {
      history.push("/historical");
    }
  }, [data, history]);

  return (
    <div className="game-wrapper">
      <Category categoryId={categoryId} />

      <Images
        userImage={userImage}
        computerImage={computerImage}
        folder={(categories as Categories[])[categoryId].folder}
      />

      <div className="settings">
        <StartGame first={first} />
        <PlayingLevel level={level} />
      </div>

      {displayBoard ? (
        <BoardUI
          userImage={userImage}
          computerImage={computerImage}
          board={board}
          winners={winners}
          folder={(categories as Categories[])[categoryId].folder}
        />
      ) : null}

      <div className="board-footer">
        <span>
          {result === GameResult.Lost
            ? MESSAGE_LOST
            : result === GameResult.Won
            ? MESSAGE_WON
            : MESSAGE_DRAW}
        </span>

        <div className="buttons">
          {!displayBoard ? (
            <button className="see-game">
              <Link to={`/historical/${id}`}>{BUTTON_SEE_GAME}</Link>
            </button>
          ) : null}

          {displayBoard ? (
            <button className="back">
              <Link to="/historical">{BUTTON_BACK}</Link>
            </button>
          ) : null}

          <button
            className="delete-game"
            onClick={() => {
              handleDeleteGame(id);
            }}
          >
            {BUTTON_DELETE_GAME}
          </button>
        </div>

        {loading && <div>{GAME_DELETING}</div>}

        {error && <div className="info">{GAME_DELETE_ERROR}</div>}
      </div>

      {!displayBoard ? <hr /> : null}
    </div>
  );
};

export default Historical;
