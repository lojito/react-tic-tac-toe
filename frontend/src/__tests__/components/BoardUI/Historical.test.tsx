import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, screen } from "@testing-library/react";
import Historical from "../../../components/BoardUI/Historical";
import { Player, SquareBoard, SquareLocation } from "../../../types";

jest.mock(
  "../../../components/Square/Historical",
  () =>
    ({ disabled, id, imagePath, className }: SquareBoard) => {
      const url =
        imagePath === "default"
          ? undefined
          : `url(${process.env.PUBLIC_URL}/images/${imagePath}.jpg)`;
      return (
        <button
          className={className}
          disabled={false}
          id={id}
          style={{ backgroundImage: url }}
        />
      );
    }
);

describe("Historical", () => {
  it("should render correctly", () => {
    const { asFragment } = render(
      <Historical
        board={[
          Player.Nobody,
          Player.Nobody,
          Player.Computer,
          Player.User,
          Player.Nobody,
          Player.Nobody,
          Player.Nobody,
          Player.Nobody,
          Player.Nobody,
        ]}
        folder="soccer"
        userImage={0}
        computerImage={1}
        winners={[]}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render with 3 winning squares", () => {
    render(
      <Historical
        board={[
          Player.Computer,
          Player.Computer,
          Player.Computer,
          Player.User,
          Player.User,
          Player.User,
          Player.Nobody,
          Player.Nobody,
          Player.Nobody,
        ]}
        folder="soccer"
        userImage={0}
        computerImage={1}
        winners={[
          SquareLocation.TopLeft,
          SquareLocation.TopCenter,
          SquareLocation.TopRight,
        ]}
      />
    );

    const squares = screen.getAllByRole("button");

    expect(squares).toHaveLength(9);
    expect(squares[SquareLocation.TopLeft]).toHaveClass("win");
    expect(squares[SquareLocation.TopCenter]).toHaveClass("win");
    expect(squares[SquareLocation.TopRight]).toHaveClass("win");
  });
});
