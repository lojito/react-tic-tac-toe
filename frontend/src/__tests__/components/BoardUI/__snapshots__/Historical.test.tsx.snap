// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Historical should render correctly 1`] = `
<DocumentFragment>
  <div
    class="board"
  >
    <button
      class=""
      id="0"
    />
    <button
      class=""
      id="1"
    />
    <button
      class=""
      id="2"
      style="background-image: url(undefined/images/soccer/1.jpg);"
    />
    <button
      class=""
      id="3"
      style="background-image: url(undefined/images/soccer/0.jpg);"
    />
    <button
      class=""
      id="4"
    />
    <button
      class=""
      id="5"
    />
    <button
      class=""
      id="6"
    />
    <button
      class=""
      id="7"
    />
    <button
      class=""
      id="8"
    />
  </div>
</DocumentFragment>
`;
